pkg/agent/controller/networkpolicy/priority.go

```diff
@@ -79,272 +98,220 @@ func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter, isSingle
+		} else {
+			prioritiesInBetween = append(prioritiesInBetween, p)
 		}
 	}
-	return insertionPoint, occupied
+	return nil, MaxUint16, prioritiesInBetween
 }
 
-// reassignPriorities re-arranges current Priority mappings to make place for the inserting Priority. It sifts
-// existing priorties up or down based on cost (how many priorities it needs to move). siftPrioritiesDown is used
-// as a tie-breaker. An error should only be returned if all the available ofPriorities in the table are occupied.
-func (pa *priorityAssigner) reassignPriorities(insertionPoint uint16, p types.Priority) (*uint16, map[uint16]uint16, func(), error) {
-	nextVacant, lastVacant := pa.getNextVacantOFPriority(insertionPoint), pa.getLastVacantOFPriority(insertionPoint)
-	switch {
-	case (insertionPoint == PolicyBottomPriority || lastVacant == nil) && nextVacant != nil:
-		return pa.siftPrioritiesUp(insertionPoint, *nextVacant, p)
-	case (insertionPoint > PolicyTopPriority || nextVacant == nil) && lastVacant != nil:
-		return pa.siftPrioritiesDown(insertionPoint-1, *lastVacant, p)
-	case nextVacant != nil && lastVacant != nil:
-		costSiftUp := *nextVacant - insertionPoint
-		costSiftDown := insertionPoint - *lastVacant - 1
+// reassignBoundaryPriorities reassigns Priorities from lowerBound / upperBound or both, to make room for
+// new Priorities to be registered. It also records all the priority updates due to the reassignment in the
+// map updates which is passed to it as parameter.
+func (pa *priorityAssigner) reassignBoundaryPriorities(lowerBound, upperBound uint16, prioritiesToRegister types.ByPriority,
+	updates map[types.Priority]*PriorityUpdate) error {
+	// gap keeps track of the vacant ofPriority thus far between lowerBound and upperBound.
+	gap := upperBound - lowerBound - 1
+	target := uint16(len(prioritiesToRegister))
+	// siftedPrioritiesLow and siftedPrioritiesHigh keeps track of Priorities that needs to be reassigned,
+	// below the lowerBound and above the upperBound, respectively.
+	var siftedPrioritiesLow, siftedPrioritiesHigh types.ByPriority
+	lowerBound, upperBound = lowerBound+1, upperBound-1
+	for gap < target {
+		lastVacant, costSiftDown, prioritiesDown := pa.getLastVacantOFPriority(lowerBound)
+		nextVacant, costSiftUp, prioritiesUp := pa.getNextVacantOFPriority(upperBound)
 		if costSiftUp < costSiftDown {
-			return pa.siftPrioritiesUp(insertionPoint, *nextVacant, p)
+			siftedPrioritiesHigh = append(siftedPrioritiesHigh, prioritiesUp...)
+			upperBound = *nextVacant
+		} else if costSiftDown < MaxUint16 {
+			siftedPrioritiesLow = append(prioritiesDown, siftedPrioritiesLow...)
+			lowerBound = *lastVacant
 		} else {
-			return pa.siftPrioritiesDown(insertionPoint-1, *lastVacant, p)
+			return fmt.Errorf("failed to push boundary priorities to either direction")
 		}
-	default:
-		return nil, map[uint16]uint16{}, nil, fmt.Errorf("no available Openflow priority left to insert priority %v", p)
+		gap++
```

Comment:
why does the gap only increase by 1 instead of `costSiftDown` / `costSiftUp`?

Reply:
This is indeed an issue... Thanks for spotting it! Because of this, one of the UT test cases is not expecting the correct outcome and I missed it. Refactored the reassign logic and updated the UT.